src/main/java/htsjdk/samtools/util/EdgingRecordAndOffset.java

```diff
@@ -0,0 +1,171 @@
+
+import static htsjdk.samtools.util.EdgingRecordAndOffset.Type.BEGIN;
+import static htsjdk.samtools.util.EdgingRecordAndOffset.Type.END;
+
+/**
+ * Holds a SAMRecord plus the zero-based offset into that SAMRecord's bases and quality scores that corresponds
+ * to the base and quality for the start of alignment block at the genomic position described by the AbstractLocusInfo.
+ * This is implementation for EdgeReadIterator, field <code>type</code> added to indicate whether object
+ * represents the start or the end of an alignment block.
+ * <p>
+ * Types BEGIN and END are used in EdgeReadIterator to distinguish starting and ending of the alignment block
+ * as for each alignment block two objects of <code>EdgingRecordAndOffset</code> are created with two different types.
+ * The main idea of using EdgeReadIterator is to process alignment block starting from locus where BEGIN type occurs,
+ * aggregate information per locus and keep it until END type occurs, then remove alignment block from consideration.
+ * 
+ * @author Darina_Nikolaeva@epam.com, EPAM Systems, Inc. <www.epam.com>
+ * 
+ */
+public class EdgingRecordAndOffset extends AbstractRecordAndOffset {
+    /**
+     * Length of alignment block of the read
+     */
+    private final int length;
+    /**
+     * A reference position to which read offset is aligned.
+     */
+    private final int refPos;
+    /**
+     * Indicates whether object represents the start or the end of an alignment block.
+     */
+    private Type type;
+    private int hash = 0;
+
+    /**
+     * For object with type END this fields holds the reference to object with type BEGIN for the read.
+     */
+    private EdgingRecordAndOffset start;
+
+    /**
+     * @param record inner SAMRecord
+     * @param offset from the start of the read
+     * @param length of alignment block
+     * @param refPos corresponding to read offset reference position
+     */
+    public EdgingRecordAndOffset(SAMRecord record, int offset, int length, int refPos, Type type) {
+        super(record, offset);
+//        validateOffset(offset, record.getBaseQualities());
+        if (length > record.getReadLength()) {
+            throw new IllegalArgumentException("Block length cannot be larger than whole read length");
+        }
+        this.type = type;
+        this.length = length;
+        this.refPos = refPos;
+    }
+
+    /**
+     * @return <code>TypedRecordAndOffset</code> that represents the start of alignment block of the read
+     * for object with type END. For object with type BEGIN will return null.
+     */
+    public EdgingRecordAndOffset getStart() {
+        return start;
+    }
+
+
+    /**
+     * Method can be called only for an object with type END.
+     * @param start <code>TypedRecordAndOffset</code> that represents the start of alignment block of the read,
+     *              must have type <code>BEGIN</code>.
+     */
+    public void setStart(EdgingRecordAndOffset start) {
```

Comment:
I've implemented this part using one abstract class EdgingRecordAndOffset and two it subclasses for BEGIN and END cases. I think code looks cleaner now and we save some memory for END records, as you suggested.